src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16LittleEndianEncoder.cs

```diff
@@ -2,102 +2,204 @@
+        /// <param name="charactersWritten">An output parameter to store the number of bytes written to <paramref name="utf16"/></param>
+        /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
+        public static bool TryEncode(uint codePoint, Span<char> utf16, int index, out int charactersWritten)
+        {
+            if (!UnicodeHelpers.IsSupportedCodePoint(codePoint))
             {
-                codePoint = default(UnicodeCodePoint);
-                encodedBytes = default(int);
-                // buffer too small
+                charactersWritten = 0;
                 return false;
             }
 
-            if (encodedBytes == 2)
+            charactersWritten = UnicodeHelpers.IsBmp(codePoint) ? 1 : 2;
+
+            if (utf16.Length - index < charactersWritten)
             {
-                unchecked
-                {
-                    buffer.Write((ushort)codePoint);
-                }
+                charactersWritten = 0;
+                return false;
             }
-            else
+
+            unchecked
             {
-                unchecked
+                if (charactersWritten == 1)
+                    utf16[index] = (char)codePoint;
+                else
                 {
-                    uint codePointValue = (uint)codePoint;
-                    uint highSurrogate = ((codePointValue - 0x010000u) >> 10) + UnicodeConstants.Utf16HighSurrogateFirstCodePoint;
-                    uint lowSurrogate = (codePointValue & MaskLow10Bits) + UnicodeConstants.Utf16LowSurrogateFirstCodePoint;
-                    buffer.Write(highSurrogate | (lowSurrogate << 16));
+                    utf16[index] = (char)(((codePoint - 0x010000u) >> 10) + Utf16HighSurrogateFirstCodePoint);
+                    utf16[index + 1] = (char)((codePoint & 0x3FF) + Utf16LowSurrogateFirstCodePoint);
                 }
             }
+
             return true;
         }
 
+        #endregion Encoding implementation
+
+        #region Surrogate helpers
+
+        private static bool IsSurrogate(uint codePoint)
+        {
+            return codePoint >= Utf16SurrogateRangeStart && codePoint <= Utf16SurrogateRangeEnd;
+        }
+
+        private static bool IsLowSurrogate(uint codePoint)
+        {
+            return codePoint >= Utf16LowSurrogateFirstCodePoint && codePoint <= Utf16LowSurrogateLastCodePoint;
+        }
+
+        private static bool IsHighSurrogate(uint codePoint)
+        {
+            return codePoint >= Utf16HighSurrogateFirstCodePoint && codePoint <= Utf16HighSurrogateLastCodePoint;
+        }
+
+        #endregion Surrogate helpers
+
+        #region Old stuff
+
         // TODO: Should we rewrite this to not use char.ConvertToUtf32 or is it fast enough?
         public static bool TryDecodeCodePointFromString(string s, int index, out UnicodeCodePoint codePoint, out int encodedChars)
         {
```

Comment:
Are we getting rid of UnicodeCodePoint completely? If so, should this be out uint?